dali/pipeline/util/bounding_box.h

```diff
@@ -20,170 +20,382 @@
+      make_string("Unexpected number of dimensions: ", ndim()));
+    for (int i = 0; i < ndim(); i++) {
+      if (point[i] < bbox_bounds_[i] || point[i] > bbox_bounds_[ndim()+i])
+        return false;
+    }
+    return true;
+  }
 
   bool Contains(float x, float y) const {
-    return x >= left_ && x <= right_ && y >= top_ && y <= bottom_;
+    DALI_ENFORCE(ndim() == 2,
+      make_string("Unexpected number of dimensions: ", ndim()));
+    return Contains(RelCoords{x, y});
   }
 
   bool Overlaps(const BoundingBox& other) const {
-    return left_ < other.right_ && right_ > other.left_ &&
-           top_ < other.bottom_ && bottom_ > other.top_;
+    for (int i = 0; i < ndim(); i++) {
+      float this_start = bbox_bounds_[i], this_end = bbox_bounds_[ndim() + i],
+            oth_start = other.bbox_bounds_[i], oth_end = other.bbox_bounds_[ndim() + i];
+      if (this_start >= oth_end || this_end <= oth_start)
+        return false;
+    }
+    return true;
   }
 
   float IntersectionOverUnion(const BoundingBox& other) const {
-    if (this->Overlaps(other)) {
-      const float intersection_area = this->ClampTo(other).area_;
-
-      return intersection_area / (area_ + other.area_ - intersection_area);
+    auto intersection_area = Intersect(other).Volume();
+    if (intersection_area == 0) {
+      return 0.0f;
     }
-    return 0.0f;
+
+    const float union_area = Volume() + other.Volume() - intersection_area;
+    return intersection_area / union_area;
   }
 
   BoundingBox RemapTo(const BoundingBox& other) const {
-    const float crop_width = other.right_ - other.left_;
-    const float crop_height = other.bottom_ - other.top_;
-
-    const float new_left =
-        (std::max(other.left_, left_) - other.left_) / crop_width;
-    const float new_top =
-        (std::max(other.top_, top_) - other.top_) / crop_height;
-    const float new_right =
-        (std::min(other.right_, right_) - other.left_) / crop_width;
-    const float new_bottom =
-        (std::min(other.bottom_, bottom_) - other.top_) / crop_height;
+    RelBounds out;
+    out.resize(ndim() * 2);
+    for (int i = 0; i < ndim(); i++) {
+      float oth_start = other.bbox_bounds_[i], start = bbox_bounds_[i];
+      float oth_end = other.bbox_bounds_[ndim()+i], end = bbox_bounds_[ndim()+i];
+      float rel_extent = oth_end - oth_start;
+      float new_start = (std::max(oth_start, start) - oth_start) / rel_extent;
+      float new_end = (std::min(oth_end, end) - oth_start) / rel_extent;
+      out[i] = clamp<float>(new_start, 0.0f, 1.0f);
+      out[ndim() + i] = clamp<float>(new_end, 0.0f, 1.0f);
+    }
+    return BoundingBox{out};
+  }
 
-    return {std::max(0.0f, std::min(new_left, 1.0f)),
-            std::max(0.0f, std::min(new_top, 1.0f)),
-            std::max(0.0f, std::min(new_right, 1.0f)),
-            std::max(0.0f, std::min(new_bottom, 1.0f))};
+  BoundingBox AxisFlip(int axis) const {
+    DALI_ENFORCE(axis >= 0 && axis < ndim(),
```

Comment:
Again, this kind of check does not belong in this class - typically, the axis would be validated at the operator's setup or even constructor. This function can merely assert.